libs/vkd3d/breadcrumbs.c

```diff
@@ -0,0 +1,648 @@
+#define NV_ENCODE_CHECKPOINT(context, counter) ((void*) ((uintptr_t)(context) + (uintptr_t)MAX_COMMAND_LISTS * (counter)))
+#define NV_CHECKPOINT_CONTEXT(ptr) ((uint32_t)((uintptr_t)(ptr) % MAX_COMMAND_LISTS))
+#define NV_CHECKPOINT_COUNTER(ptr) ((uint32_t)((uintptr_t)(ptr) / MAX_COMMAND_LISTS))
+
+static const char *vkd3d_breadcrumb_command_type_to_str(enum vkd3d_breadcrumb_command_type type)
+{
+    switch (type)
+    {
+        case VKD3D_BREADCRUMB_COMMAND_SET_TOP_MARKER:
+            return "top_marker";
+        case VKD3D_BREADCRUMB_COMMAND_SET_BOTTOM_MARKER:
+            return "bottom_marker";
+        case VKD3D_BREADCRUMB_COMMAND_SET_SHADER_HASH:
+            return "set_shader_hash";
+        case VKD3D_BREADCRUMB_COMMAND_DRAW:
+            return "draw";
+        case VKD3D_BREADCRUMB_COMMAND_DRAW_INDEXED:
+            return "draw_indexed";
+        case VKD3D_BREADCRUMB_COMMAND_DISPATCH:
+            return "dispatch";
+        case VKD3D_BREADCRUMB_COMMAND_EXECUTE_INDIRECT:
+            return "execute_indirect";
+        case VKD3D_BREADCRUMB_COMMAND_COPY:
+            return "copy";
+        case VKD3D_BREADCRUMB_COMMAND_RESOLVE:
+            return "resolve";
+        case VKD3D_BREADCRUMB_COMMAND_WBI:
+            return "wbi";
+        case VKD3D_BREADCRUMB_COMMAND_RESOLVE_QUERY:
+            return "resolve_query";
+        case VKD3D_BREADCRUMB_COMMAND_GATHER_VIRTUAL_QUERY:
+            return "gather_virtual_query";
+        case VKD3D_BREADCRUMB_COMMAND_BUILD_RTAS:
+            return "build_rtas";
+        case VKD3D_BREADCRUMB_COMMAND_COPY_RTAS:
+            return "copy_rtas";
+        case VKD3D_BREADCRUMB_COMMAND_EMIT_RTAS_POSTBUILD:
+            return "emit_rtas_postbuild";
+        case VKD3D_BREADCRUMB_COMMAND_TRACE_RAYS:
+            return "trace_rays";
+        case VKD3D_BREADCRUMB_COMMAND_BARRIER:
+            return "barrier";
+        case VKD3D_BREADCRUMB_COMMAND_AUX32:
+            return "aux32";
+        case VKD3D_BREADCRUMB_COMMAND_AUX64:
+            return "aux64";
+        case VKD3D_BREADCRUMB_COMMAND_VBO:
+            return "vbo";
+        case VKD3D_BREADCRUMB_COMMAND_IBO:
+            return "ibo";
+        case VKD3D_BREADCRUMB_COMMAND_ROOT_DESC:
+            return "root_desc";
+        case VKD3D_BREADCRUMB_COMMAND_ROOT_CONST:
+            return "root_const";
+
+        default:
+            return "?";
+    }
+}
+
+HRESULT vkd3d_breadcrumb_tracer_init(struct vkd3d_breadcrumb_tracer *tracer, struct d3d12_device *device)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = &device->vk_procs;
+    D3D12_HEAP_PROPERTIES heap_properties;
+    D3D12_RESOURCE_DESC1 resource_desc;
+    HRESULT hr;
+    int rc;
+
+    memset(tracer, 0, sizeof(*tracer));
+
+    if ((rc = pthread_mutex_init(&tracer->lock, NULL)))
+        return hresult_from_errno(rc);
+
+    if (device->device_info.device_coherent_memory_features_amd.deviceCoherentMemory &&
```

Comment:
Is it possible to make `VK_AMD_device_coherent_memory` optional? In Turnip we don't have it implemented. On the other hand our memory by default is uncached | coherent, but writes to it from GPU may still have to be flushed from internal caches (the AMD extension seem to imply that any write to such memory bypasses all caches).

Reply:
> Can you guarantee that the marker write is coherent/uncached?

We have two means of writing the marker:
- For top of the pipe marker - via `CP_MEM_WRITE` which goes into a cache but we also dirty it so the next draw/blit would flush it.
- For others - through `CP_EVENT_WRITE` which isn't cached.

And the underlying memory itself is uncached and coherent.

Reply:
Fixed.